### DISCUSSION TITLE
用小程序单元测试芋头

### DISCUSSION THREAD

**Discussion by YoavAtzmon** [NervJS/taro#11910] [Q & A]
你好

我们正在用 Taro 结合 wixin（小程序）开发一个移动应用程序，

当我们只对干净的 Taro 组件执行单元测试时，它是有效的：

![image](https://user-images.githubusercontent.com/90871786/172791704-5c99bb8e-7e9e-4711-8c4d-83ad6ae66c24.png)

![image](https://user-images.githubusercontent.com/90871786/172790349-64c2291f-481b-4316-9383-a6c37f72384e.png)

但是当我们使用来自 wixin(miniprogram) 的导入时：

![image](https://user-images.githubusercontent.com/90871786/172790620-3feb635c-2843-43b5-b609-fd0c54ad1562.png)

/taro/src/utils/globalData : 

![image](https://user-images.githubusercontent.com/90871786/172790656-e19bd266-d7c0-4f89-9851-43c94d3505bc.png)

/mina/src/utils/storage/globalData : 

![image](https://user-images.githubusercontent.com/90871786/172790877-9f07c91c-150b-40c2-98ae-2ac3bc580b1d.png)

我们得到下一个错误：

![image](https://user-images.githubusercontent.com/90871786/172791343-89c63a55-3cb1-48d9-8cb9-45fde4e29fce.png)

在多次尝试解决此问题失败后，我们决定与您联系以找到解决此问题的方法。

这是我们的 Taro package.json 环境：

{
    "name": "dior-taro",
    "version": "1.0.0",
    "private": true,
    "description": "taro app",
    "templateInfo": {
        "name": "default",
        "typescript": true,
        "css": "sass"
    },
    "scripts": {
        "build": "taro build --type weapp --blended",
        "start": "taro build --type weapp --blended",
        "dev": "taro build --type weapp --blended --watch",
        "test": "jest"
    },
    "browserslist": [
        "last 3 versions",
        "Android >= 4.1",
        "ios >= 8"
    ],
    "author": "",
    "dependencies": {
        "@babel/runtime": "^7.7.7",
        "@linaria/babel-preset": "^3.0.0-beta.15",
        "@linaria/core": "^3.0.0-beta.15",
        "@linaria/react": "^3.0.0-beta.15",
        "@linaria/shaker": "^3.0.0-beta.15",
        "@reduxjs/toolkit": "^1.8.0",
        "@tarojs/cli": "^3.3.16",
        "@tarojs/components": "3.3.9",
        "@tarojs/mobx": "^2.2.10",
        "@tarojs/react": "3.3.12",
        "@tarojs/runtime": "3.3.12",
        "@tarojs/taro": "3.3.12",
        "@testing-library/jest-dom": "^5.16.4",
        "@testing-library/react": "^12.1.2",
        "@testing-library/user-event": "^14.2.0",
        "@types/enzyme": "^3.10.12",
        "@types/jest": "^27.5.1",
        "@types/jsdom": "^16.2.14",
        "@wojtekmaj/enzyme-adapter-react-17": "^0.6.7",
        "babel-jest": "^28.1.0",
        "chai": "^4.3.6",
        "dayjs": "^1.10.8",
        "enzyme": "^3.11.0",
        "jest": "^28.1.0",
        "react": "^17.0.0",
        "react-dom": "^17.0.0",
        "react-redux": "^7.2.6",
        "react-swipe-to-reveal-actions": "^1.0.3",
        "react-swipeable-views": "^0.14.0",
        "redux": "^4.1.2",
        "redux-persist": "^6.0.0",
        "taro-calendar-customizable": "^2.1.2",
        "taro-skeleton": "^2.0.4",
        "taro-skeleton-component": "^0.0.2",
        "ts-jest": "^28.0.3"
    },
    "devDependencies": {
        "@babel/core": "^7.8.0",
        "@tarojs/mini-runner": "3.3.9",
        "@tarojs/plugin-platform-weapp-qy": "0.0.7",
        "@tarojs/webpack-runner": "3.3.9",
        "@types/react": "^17.0.2",
        "@types/webpack-env": "^1.13.6",
        "@typescript-eslint/eslint-plugin": "^4.15.1",
        "@typescript-eslint/parser": "^4.15.1",
        "babel-preset-taro": "3.3.9",
        "cheerio": "^1.0.0-rc.3",
        "eslint": "^6.8.0",
        "eslint-config-taro": "3.3.9",
        "eslint-plugin-import": "^2.12.0",
        "eslint-plugin-react": "^7.8.2",
        "eslint-plugin-react-hooks": "^4.2.0",
        "jsdom": "19.0.0",
        "jsdom-global": "3.0.2",
        "linaria": "^2.0.2",
        "stylelint": "9.3.0",
        "taro-testing-library": "^1.1.0",
        "typescript": "^4.1.0",
        "vitest": "^0.14.1"
    }
}

这是我们的 Taro jest.config.js 环境：

module.exports = {
    verbose: true,
    moduleNameMapper: {
        '@tarojs/components': '@tarojs/components/dist-h5/react',
        '^.+\\.(css|scss|less)$': '<rootDir>/style-mock.js'
    },
    transform: {
        '^.+\\.(js|jsx|ts|tsx)$': '<rootDir>/node_modules/babel-jest',
        '^.+\\.esm.js?$': 'ts-jest',
    },
    rootDir: __dirname,
    setupFiles: ['<rootDir>/tests/setupTests'],
    moduleFileExtensions: ['js', 'jsx', 'ts', 'tsx', 'json'],
    transformIgnorePatterns: ['<rootDir>/node_modules/(?!@taro)', '^.+\\.(css|sass|scss|less)$'],
};

谢谢

**Comment by Chen-jj** [comment]
@YoavAtzmon mock 一下 `getApp` 等小程序特有的 API 应该就 ok
